Searching and adding new creators

awead opened the pull request:
This adds the bulk of the features for adding new creators to works. There are still outstanding accessibility issues and the requirements for metadata need to be changed. These can be addressed in forthcoming tickets.

Fixes #244

awead commented:
@rschenk I added some bit for a11y and styled the search input like the other inputs, but it may need some work.

* the input doesn't look quite right
* card heading text needs to be adjusted due to the icon size

![Screen Shot 2020-09-24 at 12 06 15 PM](https://user-images.githubusercontent.com/312085/94170759-73b76080-fe5e-11ea-96ef-0fc095e098a8.png)

rschenk commented on app/javascript/controllers/actors_controller.js:
```diff
@@ -0,0 +1,40 @@
+  // This avoids having to have a separate .js.erb partial in a typical Rails AJAX setup.
+  success (event) {
+    const [data, status, xhr] = event.detail // eslint-disable-line no-unused-vars
+    document.getElementsByClassName('modal-content').item(0).innerHTML = data.body.innerHTML
```
Even though there's only one modal in the layout (currently!), it would be more robust and more idiomatic javascript if the modal-content element had an ID, and this line here used `getElementById`. It's a good habit to try to use IDs and not classes whenever searching for elements in js (sometimes lookup by class is totally warranted, but much of the time it's not a good idea)

rschenk commented on app/javascript/controllers/actors_controller.js:
```diff
@@ -0,0 +1,40 @@
+  append (event) {
+    event.preventDefault()
+    const data = { actor_id: event.target.dataset.id }
+    data.index = document.getElementsByClassName('nested-fields').length
```
There might be more than one set of nested-fields on the page in the future. In Christie's mockups, Creators, Collaborators, and Acknowledgements are all on the same tab (and might have nested fields). Probably a good idea to scope this to within `#creator_aliases`

rschenk commented on app/javascript/controllers/autocomplete_controller.js:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  appendSuggestion (suggestion) {
+    suggestion.index = document.getElementsByClassName('nested-fields').length
```
Ditto with `actors_controller.js` Scope this `.nested-fields` to a particular element, in case we use .nested-fields class for multiple components on the page.

rschenk commented on app/javascript/controllers/actors_controller.js:
```diff
@@ -0,0 +1,40 @@
+    })
+      .then(function (response) {
+        document
+          .getElementById('creator_aliases')
```
Might be a cool idea to pass the `'creator_aliases'` id in as a data-attribute and really decouple this logic from the html

rschenk commented on app/javascript/controllers/autocomplete_controller.js:
```diff
@@ -0,0 +1,61 @@
+    })
+      .then(function (response) {
+        document
+          .getElementById('creator_aliases')
```
Might be a cool idea to pass the `'creator_aliases'` id in as a data-attribute and really decouple this logic from the html

rschenk commented on app/javascript/controllers/stimulus_modules.js:
```diff
@@ -0,0 +1,23 @@
+
+// Move to the last creator added to the form
+export function jumpToCreator () {
+  const nodes = document.getElementsByClassName('card-body')
```
This seems really brittle, because this is a bootstrap class used for styling. If we decided to restyle these, it would break this javascript. Best bet here would be to create a new class for creator fields, and if you want to go full BEM style use a "js" prefix so someone coming in to tweak the html would know to think twice before changing it, like  `js-creator-fields`

rschenk reviewed (changes requested):
I had a couple really specific technical comments with the javascript, but I also have a couple UX feedbacks too... 

* I wonder if we need another method to pop open the "new actor" modal, besides completely empty search results. It's possible that I could search for someone who is not in the list, and yet still have search results—in that scenario I can't add them. Perhaps there should always be a "add someone not in this list" at the very bottom of the auto complete?

* If the creator search field loses focus, I can't get the list to pop open again. If I do a search for "Wead", click away from the field so that it loses focus, then focus it again, the results list does not reappear. I have to change the text, like delete the "d" and add it again

* In the New Creator modal, I'm not sure if the two-step process of making them confirm the Actor, then click Add Creator to Work on the second page is useful. I can't see any reason why a user _wouldn't_ want to add them to the work, since this is why they opened the dialog to begin with. I think making them confirm will be confusing and lead to errors. I reckon if the New Creator form submits correctly without validations, it should just be inserted into the main form. That is the desired action after all.

* We really need to beef up our validations and constraints on the Actor model/table. For example, I can add multiple people with the same PSU access ID. Yikes.

* Despite not totally following Christie's design for this page, I do think it'd be nice to implement most of the styling. I will attach the drawing as a separate comment.

rschenk commented:
Even if we aren't going completely with Christie's UX design with the modals, I do think it'd be nice to update our forms to use some/most of her visual styling

![contributors2](https://user-images.githubusercontent.com/14730/94187364-d7995380-fe75-11ea-9116-960915271dab.png)

![addCreators](https://user-images.githubusercontent.com/14730/94187378-dc5e0780-fe75-11ea-8329-0943340bb9ed.png)

awead commented:
> I had a couple really specific technical comments with the javascript, but I also have a couple UX feedbacks too...
> 
>     * I wonder if we need another method to pop open the "new actor" modal, besides completely empty search results. It's possible that I could search for someone who is not in the list, and yet still have search results—in that scenario I can't add them. Perhaps there should always be a "add someone not in this list" at the very bottom of the auto complete?
> 

fixed!

>     * If the creator search field loses focus, I can't get the list to pop open again. If I do a search for "Wead", click away from the field so that it loses focus, then focus it again, the results list does not reappear. I have to change the text, like delete the "d" and add it again
> 

fixed!

>     * In the New Creator modal, I'm not sure if the two-step process of making them confirm the Actor, then click Add Creator to Work on the second page is useful. I can't see any reason why a user _wouldn't_ want to add them to the work, since this is why they opened the dialog to begin with. I think making them confirm will be confusing and lead to errors. I reckon if the New Creator form submits correctly without validations, it should just be inserted into the main form. That is the desired action after all.
> 

fixed. New actors are added directly to the form.

>     * We really need to beef up our validations and constraints on the Actor model/table. For example, I can add multiple people with the same PSU access ID. Yikes.
> 

ticket separately?

>     * Despite not totally following Christie's design for this page, I do think it'd be nice to implement most of the styling. I will attach the drawing as a separate comment.

amended!
